Groovy ?. operator on null File objects throws exception

### Current Behavior

Using the elvis operator on a File reference in the Groovy DSL causes an exception if the reference is null:

A problem occurred evaluating root project 'gradle9test'.
> Parameter specified as non-null is null: method org.gradle.internal.cc.impl.InstrumentedInputAccessListener.fileSystemEntryObserved, parameter file
...
Caused by: java.lang.NullPointerException: Parameter specified as non-null is null: method org.gradle.internal.cc.impl.InstrumentedInputAccessListener.fileSystemEntryObserved, parameter file
        at org.gradle.internal.cc.impl.InstrumentedInputAccessListener.fileSystemEntryObserved(InstrumentedInputAccessListener.kt)
        at org.gradle.internal.classpath.Instrumented.fileSystemEntryObserved(Instrumented.java:319)
        at org.gradle.internal.classpath.Instrumented.fileIsDirectory(Instrumented.java:333)
        at org.gradle.internal.classpath.declarations.FileInterceptorsDeclaration.intercept_isDirectory(FileInterceptorsDeclaration.java:89)
        at org.gradle.internal.classpath.generated.InterceptorDeclaration_ConfigCacheGroovyInterceptors$IsDirectoryCallInterceptor.intercept(InterceptorDeclaration_ConfigCacheGroovyInterceptors.java:166)
        at org.gradle.internal.instrumentation.api.groovybytecode.AbstractCallInterceptor.interceptMethodHandle(AbstractCallInterceptor.java:61)

`
The example project attached works in Gradle 8.14.3 but throws an exception in 9.0.0.

### Expected Behavior

No exception should be thrown.

### Context (optional)

_No response_

### Self-contained Reproducer Project

[gradle9test.zip](https://github.com/user-attachments/files/22423618/gradle9test.zip)

### Gradle version

9.0.0

### Build scan URL (optional)

_No response_

### Your Environment (optional)

macOS 14.7.8, openjdk version "17.0.10" 2024-01-16

--- a/platforms/core-runtime/internal-instrumentation-api/src/main/java/org/gradle/internal/instrumentation/api/groovybytecode/AbstractCallInterceptor.java
+++ b/platforms/core-runtime/internal-instrumentation-api/src/main/java/org/gradle/internal/instrumentation/api/groovybytecode/AbstractCallInterceptor.java
@@ -57,6 +57,11 @@ public abstract class AbstractCallInterceptor implements CallInterceptor {
     @Nullable
     @SuppressWarnings("unused")
     private Object interceptMethodHandle(MethodHandle original, int flags, String consumer, Object[] args) throws Throwable {
+        boolean isSafeNavigation = (flags & IndyInterface.SAFE_NAVIGATION) != 0;
+        if (isSafeNavigation && InvocationUtils.unwrap(args[0]) == null) {
+            // Skip interception for safe navigation calls on null receiver
+            return original.invokeExact(args);
+        }
         boolean isSpread = (flags & IndyInterface.SPREAD_CALL) != 0;
         return intercept(new MethodHandleInvocation(original, args, isSpread), consumer);
     }

--- a/subprojects/core/src/test/groovy/org/gradle/internal/classpath/GroovyDynamicDispatchingInterceptingTest.groovy
+++ b/subprojects/core/src/test/groovy/org/gradle/internal/classpath/GroovyDynamicDispatchingInterceptingTest.groovy
@@ -18,6 +18,7 @@ package org.gradle.internal.classpath
 
 import org.gradle.internal.classpath.intercept.JvmBytecodeInterceptorFactoryProvider
 import org.gradle.internal.metaobject.BeanDynamicObject
+import spock.lang.Issue
 
 import java.lang.reflect.Constructor
 import java.util.function.Predicate
@@ -202,5 +203,27 @@ class GroovyDynamicDispatchingInterceptingTest extends AbstractCallInterceptionT
 
         then:
         receiver.intercepted == "test()"
+    }
+
+    @Issue("https://github.com/gradle/gradle/issues/35049")
+    def 'safe navigation works with intercepted calls'() {
+        given:
+        def receiver = new InterceptorTestReceiver()
+        def transformedClosure = instrumentedClasses.instrumentedClosure { delegate?.test() }
+        transformedClosure.delegate = receiver
+
+        when: "typical non-null call"
+        transformedClosure()
+
+        then:
+        receiver.intercepted == "test()"
+
+        when: "null delegate that should be safe"
+        receiver.intercepted = null
+        transformedClosure.delegate = null
+        transformedClosure()
+
+        then:
+        receiver.intercepted == null
     }
 }

